Remove NotImplemented handling from the ufunc machinery (almost)

njsmith opened the pull request:
We've long had consensus that returning `NotImplemented` from ufuncs was a bad and ugly hack, since it violates layering -- `NotImplemented` is the responsibility of `__op__` methods, not ufuncs. And one of the things that came out of the discussion in gh-5844 is that the `NotImplemented` handling inside the ufunc machinery was actually mostly useless and broken (see the first post on that report for analysis).

Of course, it turns out it's not _quite_ that simple -- there were some places, esp. in the trainwreck that is `array_richcompare`, which were depending on the current behavior.

This patch series fixes the main pieces of code that used to rely on this behavior, and then removes as much of the `NotImplemented` handling as it can. And for what isn't removed, it adds deprecation or future warnings.

In particular, there are a bunch of cases where `array_richcompare` returns bad and wrong results, like where you can compare two arrays (which should produce a boolean array), some failure occurs, and then the error gets thrown away and we end up returning a boolean scalar. Some of these were deprecated in 9b8f6c72, but this deprecates the rest. Probably the most controversial part of this is that this patch deprecates scalar ordering comparisons between unlike types, like `np.float64(1) < "foo"`, which are already an error on py3 but have traditionally returned some semi-arbitrary value on py2. To convince you that this is broken and we should stop supporting it, check this out:

```
# Note: the name of this class affects the results below
In [1]: class quux(object):
   ...:     pass
   ...: 

In [2]: q = quux()

In [3]: np.string_("foo") < q
Out[3]: True

In [4]: np.asarray(np.string_("foo")) < q
Out[4]: False
```

For a full analysis of how we handle all the legacy comparison cases, see the long comment that starts here: https://github.com/numpy/numpy/blob/19b829a660a6adc8118dc427e1a7e4ba6946ccf7/numpy/core/src/umath/ufunc_object.c#L867

NB this PR is probably easier to review if you read the commits one at a time in order, instead of jumping straight to the big diff.

njsmith committed:
MAINT: move the special case for string comparison before the regular case

The ndarray richcompare function has a special case for handling
string dtypes (which currently cannot be handled by
ufuncs). Traditionally this was handled by the ufuncs returning
NotImplemented, and then falling through to a special case. By moving
the special case to the top of the richcompare function, it becomes
unnecessary for the ufuncs to return NotImplemented in this case.

njsmith committed:
MAINT: move the special case for void comparison before the regular case

The ndarray richcompare function has special case code for handling
void dtypes (esp. structured dtypes), since there are no ufuncs for
this. Previously, we would attempt to call the relevant
ufunc (e.g. np.equal), and then when this failed (as signaled by the
ufunc returning NotImplemented), we would fall back on the special
case code. This commit moves the special case code to before the
regular code, so that it no longer requires ufuncs to return
NotImplemented.

Technically, it is possible to define ufunc loops for void dtypes
using PyUFunc_RegisterLoopForDescr, so technically I think this commit
changes behaviour: if someone had registered a ufunc loop for one of
these operations, then previously it might have been found and
pre-empted the special case fallback code; now, we use the
special-case code without even checking for any ufunc. But the only
possible use of this functionality would have been if someone wanted
to redefine what == or != meant for a particular structured dtype --
like, they decided that equality for 2-tuples of float32's should be
different from the obvious thing. This does not seem like an important
capability to preserve.

There were also several cases here where on error, an array comparison
would return a scalar instead of raising. This is supposedly
deprecated, but there were call paths that did this that had no
deprecation warning. I added those warnings.

njsmith committed:
MAINT: add __array_priority__ special case to masked array binary ops

ndarray special methods like __add__ have a special case where if the
right argument is not an ndarray or subclass, and it has higher
__array_priority__ than the left argument, then we return
NotImplemented and let the right argument handle the operation.

ufuncs have traditionally had a similar but different special case,
where if it's a 2 input - 1 output ufunc, and the right argument is
not an ndarray (exactly, subclasses don't count), and when converted
to an ndarray ends up as an object array (presumably b/c it doesn't
have a meaningful coercion route, though who knows), and it has a
higher __array_priority__ than the left argument AND it has a
__r<operation>__ attribute, then they return NotImplemented.

In practice this latter special case is not used by regular ndarrays,
b/c anytime it would need to be triggered, the former special case
triggers first and the ufunc is never called. However, numpy.ma did
not have the former special case, and was thus relying on the ufunc
special case. This commit adds the special case to the numpy.ma
special methods directly, so that they no longer depend on the quirky
ufunc behaviour.

It also cleans up the relevant test to things that actually should be
true in general, instead of just testing some implementation details.

njsmith committed:
MAINT: Remove NotImplemented handling from ufuncs (almost)

This was redundant/broken anyway.

See gh-5844 for discussion. See the massive comment added to
ufunc_object.c:get_ufunc_arguments for a discussion of the deprecation
strategy here -- it turns out that array_richcompare is such a
disaster zone that we can't quite wholly eliminate NotImplemented
quite yet. But this removes most NotImplementeds, and lays the
groundwork for eliminating the rest in a release or two.

njsmith commented:
Added release notes.

mhvk commented:
@njsmith - I looked at most commits (though void comparison is beyond me), and like the general idea. One small worry: with your pre-ufunc testing, I think the regular numeric cases are slowed down more than they are now. Though from a quick look, e.g., `PyArray_ISSTRING` would seem to be a pretty fast function, so probably this is not an issue. However,  I don't understand the C code well enough to be sure, so thought I would mention it.

mhvk commented:
More importantly: Are you sure about what `==` and `!=` should do? I thought the python rule was that `==` always returns `False` if the objects cannot be compared with each other. Implicit in returning a full array of `False` would seem to be that an actual comparison is done for each element. In this indeed intended? It implies things that seem odd to me; e.g., should one expect that:
```
np.arange(3) == [0., 'bla', 2.]
# -> array([ True, False,  True], dtype=bool)
```
(i.e., just as what would happen if I compared with `[0., 3., 2.]`). 

Of course, while this is different from current behaviour and thus breaks backward compatibility, it is internally consistent, and the present deprecation warnings suggest it was discussed before. If so, no problem. But if the above would not be expected future behaviour, then maybe fcous this PR just on the removal of improper usage of `NotImplemented` inside ufuncs, leaving the updated warnings/deprecations for a different PR?

njsmith commented:
@mhvk: Thanks for the comment (that now seems to have disappeared into the aether somewhere) about the reversed operators in np.ma -- it of course was left over from an earlier version that tried to duplicate all the details of the old ufunc special case, before I said screw it, any `__array_priority__` special case that's good enough for ndarray is surely also good enough for np.ma. Fixed.

Regarding the speed issue: yeah, there are only two extra checks that are done in the common case: one is `if (PyArray_ISSTRING(self)) ...`, which is a macro that expands to `if (self->descr->type_num == NPY_STRING || self->descr->type_num == NPY_UNICODE)`, so incredibly cheap, and the other is `if (PyArray_TYPE(self) == NPY_VOID)`, which has a similar expansion. So I don't think we need to worry about this.

Regarding the proper behavior of `==` and `!=` for arrays: I think it's pretty well-established that the proper behavior for these is to perform an elementwise comparison. (This is indeed different from Python's default assumptions, but Python doesn't have arrays, and this is a place where we intentionally deviate.) My feeling is that the equivalent of Python's rule that all objects are comparable for equality is that all array elements should be comparable for equality, i.e. if you have two uncomparable dtypes then you should get an array of `False`s. I was actually surprised by your example, because I thought that it would in fact return `[True, False, True]`... consider this, which works right now:
```
In [3]: np.arange(3) == np.asarray([0.0, "blah", 2.0], dtype=object)
Out[3]: array([ True, False,  True], dtype=bool)
```
The reason you get a different result there is because it turns out that `asarray` applied to that particular list does not return an object array the way I expected, but instead produces this terribleness:
```
In [4]: np.asarray([0.0, "blah", 2.0])
Out[4]: 
array(['0.0', 'blah', '2.0'], 
      dtype='|S32')
```
(Not only is doing string coercion here broken, but... why 32 byte strings? Though at least that's a power of two -- OTOH if you make the numbers integers instead of floats, you get 21 byte strings. WTF? It's just fractally weird.)

The reason the deprecations etc. are here in this PR is b/c there's no reasonable way to produce similar behaviour to what we have now without having that one `NotImplemented` deep inside the ufunc machinery, so the only way to get rid of it eventually is to add warnings around it while leaving it there for now. The alternative would be to just make all these cases into errors immediately, but that might make it harder to get this merged and I'd rather make some forward progress :-).

njsmith commented:
Fixed py3 test failures (let's see if Travis agrees).

pv commented:
The getpriority checks inside ufunc (which are removed here) are
responsible for making __array_priority__ be respected for inplace ops.
Some copypasting of the getpriority check in number.c is required to
make it work again.
.
Note also that the check in number.c has PyArray_Check(other), but the
check in the ufunc has PyArray_CheckExact(other), so without me testing
it it looks like there's a behavior change.

pv commented:
To be able to do this refactoring safely, it's probably necessary to
exercise the various branches in the code, and to have additional tests.
It seems as if to cover the different branches, checks for operations
involving (i) ndarray & ndarray subclass, (ii) ndarray & non-ndarray
object, (iii) ndarray subclass & different ndarray subclass, (vi)
ndarray subclass & other object.

njsmith commented:
I guess I don't know what it even means to respect `__array_priority__` for inplace ops. There is no `__iradd__` (nor would it make a lot of sense to have one), so... can you elaborate? What *does* happen if you return `NotImplemented` from `__iadd__`? Which code exactly are you saying needs updating?

Re: `PyArray_Check` versus `PyArray_CheckExact` in the number.c and ufunc `__array_priority__` special cases respectively: `PyArray_CheckExact` is a stricter check, so removing it from the ufunc code has no effect on any code that goes through the number.c call path. (Anything that might possibly hit the ufunc check will have already hit the number.c check and bailed out before calling the ufunc.) AFAICT the only code that cared about the ufunc check but didn't go through the number.c check is the `np.ma` special methods, so it's the only code affected by this, and I fixed that by copying the number.c check into np.ma. So this is a behavioural change -- it makes `np.ma` become more consistent with `ndarray`. If you're worried, we could instead copy the old ufunc check into `np.ma`... I suspect this would be a little more broken, but I don't care that much either way, because `np.ma` is always fragile and a bit broken in corner cases...

If there are other paths besides np.ma that depended on the ufunc `__array_priority__` check that I missed, then those indeed might be broken. Do you see any? (I guess you're saying the in-place ops are such a path?)

I believe the np.ma and np.matrix tests cover at least (i) and (iv) on your list. (FYI your email originally said "(vi)" instead of "(iv)" but I went ahead and edited this on github for clarity -- hopefully this is okay.) Basically you just want checks that show that dispatch does get delegated given an appropriate `__array_priority__` value, that should pass identically on both 1.9 and trunk?

pv commented:
If you return NotImplemented from __iadd__, Python does x = x + y.
*GenericInplace* in number.c is probably a good place to insert the
array priority check.
.
The check in number.c is weaker due to negation, but indeed the one in
ufunc is only applies to object dtype (after cast). So the behavior
indeed is fairly broken in master, array_priority doesn't affect
ordering between different ndarray subclasses :(
.
Yes, I'd prefer to have explicit tests for the different cases. They are
perhaps partly covered by matrix/maskedarray tests, but it is easier to
see if a test coverage is complete if the tests are in a single place
targeted at a single things.
.
Apart from the inplace op issues, I don't immediately (don't have time
to take a good look right now, however) see other backward incompat
issues. I would guess that because array_priority works so
inconsistently, it's not very common to have code that explicitly relies
on this. For backward compat, testing against known modules dealing with
ndarray binops, scipy.sparse, cvxopt, etc. can probably smoke up
remaining issues, if any.

pv commented:
Ok, I see that the inplace op override only applies when lhs is an
object array. Fairly unlikely anyone depends on this then.

mhvk commented:
@njsmith - OK, so it is the present that is broken!  Doing the comparison by array element is indeed fine, if done consistently.  Still am somewhat amused by this, but it is consistent:
```
np.arange(2, 5) * np.array([0., 'bla', 2.], dtype='O')
#-> array([0.0, 'blablabla', 8.0], dtype=object)
```

mhvk commented:
I checked your branch on astropy and there is one test that fails (but passes on current master). The equivalent non-astropy test is below; this suggests `__index__` is no longer properly used (which in turn suggests `list.__rmul__` is not tried -- `NotImplemented` hitting again?): 

```
np.array([4]) * ['a', 'b', 'c']
# on master: -> ['a', 'b', 'c', 'a', 'b', 'c', 'a', 'b', 'c', 'a', 'b', 'c']
# on your branch:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-99d078c7927a> in <module>()
----> 1 np.array([4]) * ['a', 'b', 'c']

TypeError: ufunc 'multiply' did not contain a loop with signature matching types dtype('<U21') dtype('<U21') dtype('<U21')
```

charris commented:
Tagging this for 1.10 release. This functionality needs to be finalized so that it can be used in development.

mhvk commented:
Modulo the odd failure I mention above, which may just need a "we don't care" decision, this looks good.

charris commented:
While we are at this, how should we handle `dot`, `__matmul__` and `__rmatmul__`. None of those are currently ufuncs. Has anyone tried overriding `dot`?

njsmith commented:
@charris: `__numpy_ufunc__` has a dirty secret: `np.dot` pretends to be a ufunc and calls `__numpy_ufunc__` if present, even though `dot` is not actually a ufunc. This is a sin, of course, which we can expiate by eventually turning `dot` into a ufunc.

charris commented:
Yeah, I'm aware of that. I have a current version of matmul (for the `@` operator) that does the same thing. I'm a bit more concerned with the methods as they have both left and right versions like the other builtin operators.

I've also done work for matmul as a ufunc, but am not pursueing that at the moment.

njsmith commented:
@mhvk: Ughhhhh wow that's terrible. Also, it turns out that the behavior for these cases depends crucially on some long-standing bugs in Python's own dispatch rules, woohoo:
```
# Using numpy 1.9:

# ndarray.__mul__ calls np.multiply which returns NotImplemented b/c strings are hard
# This causes Python to call list.__rmul__, which calls ndarray.__index__ to convert the
# array to an integer (see sequence_repeat() in abstract.c in the CPython source)
In [2]: np.array(2) * ["a", "b"]
Out[2]: ['a', 'b', 'a', 'b']

# Obviously, you also get the same result if you just call list.__mul__ to start with
In [3]: ["a", "b"] * np.array(2)
Out[3]: ['a', 'b', 'a', 'b']

# Here's another case where we just call list.__mul__, and ndarray methods are never
# called, so we can't possibly get anything differe-- wait WTF?
In [4]: [1, 2] * np.array(2)
Out[4]: array([2, 4])

In [5]: np.array(2) * [1, 2]
Out[5]: array([2, 4])
```

What's going on is this: in the CPython type struct, there are actually two different methods that both map to `__mul__`: `tp_as_number->nb_multiply` and `tp_as_sequence->sq_repeat`. When you do `a * b`, what actually happens is:
* Call `a->ob_class->tp_as_number->nb_multiply(a, b)`. If this doesn't exist or returns `NotImplemented`...
* Call `b->ob_class->tp_as_number->nb_multiply(a, b)`. If this doesn't exist or returns `NotImplemented`...
* Call `a->ob_class->tp_as_sequence->sq_repeat(a, b)`. If this doesn't exist or returns `NotImplemented`...
* Call `b->ob_class->tp_as_sequence->sq_repeat(a, b)`. If this doesn't exist or returns `NotImplemented`...
* raise `TypeError`.
(See `PyNumber_Multiply` in `abstract.c`.)

So notice that this means that we actually try `a`, then `b` reversed, then `a` *again*, and then `b` reversed *again*. `list`'s multiplication method is part of `tp_as_sequence`, so it only gets checked on the second pass, meaning that when you do `list * ndarray`, `ndarray.__rmul__` is actually tried *before* `list.__mul__`, even though neither is a subclass of the other.

Now, what should we do.

I take it as given that we eventually want `np.array(2) * ["a", "b"]` and `np.array(2) * [1, 2]` to eventually either *both* be elementwise or *both* not be, i.e., strings should be treated like a regular dtype not some weird magical thing that breaks through the abstraction boundary between ndarray and dtype. I also take it as given that `np.array(2) * np.asarray([1, 2])` is always going to be elementwise, and that we'd prefer for `np.array(2) * [1, 2]` to be the same, following the usual rule that lists get coerced to arrays. This would argue that `np.array(2) * ["a", "b"]` should either raise an error or maybe return `np.array(["aa", "bb"])`, but certainly not `["a", "b", "a", "b"]`. So, the current behaviour in this PR is correct. (Except maybe if someone thinks we need to do a deprecation cycle before changing this, but given how specific and weird the behaviour is, and how tricky it would be to manage the deprecation, I would really rather not? Please speak up if you disagree, but otherwise I'm just going to assume that if we decide to make this an error, then it's okay to make that happen directly without a transition period.)

There is one reason this gives me pause, though, which is this: it's easy to get the semantics we want in CPython -- here's a [python-dev post from 2005](https://mail.python.org/pipermail/python-dev/2005-October/056905.html) mentioning this weirdness, and we're a prominent project which is already actively depending on it, so I guess it's not going anywhere. But AFAIK it *ought* not to be possible to do this, given a strict reading of the Python language semantics -- which means that it might not be possible to do this on other Python interpreters, like PyPy. So I guess I'll send a note to python-dev asking about that...

[Can I just say I will be very glad when we reach the point where our dispatch rules are settled and simple and we can forget about all these details again.]

njsmith commented:
@charris: I guess in a perfect world (according to me, see #5844), `__matmul__` and `__rmatmul__` and `__imatmul__` would all just call `np.dot` with appropriate arguments, and that would be a ufunc, which would call `__numpy_ufunc__` if defined. Obviously there are multiple pieces in that list that we can't do immediately though. In the short term, I guess our only reasonable option is to either define a `matmul` ufunc or (if this is too much work) then define a `matmul` function which is like `dot`, and pretends to be a ufunc WRT `__numpy_ufunc__` dispatch. And then we can use the same rule for `__matmul__` and friends dispatching to `matmul` that we use for other binops (whatever that ends up being -- #5844 is the discussion of that).

njsmith commented:
@pv: I still haven't had a chance to dig into the details of how the inplace ops are currently working, and I'm not quite sure what you mean from what you wrote (which I guess is just because you weren't sure yet at the time either :-)). Can you easily re-say what exactly you think this PR needs to do WRT in-place ops?

njsmith committed:
MAINT: give more details on assert_deprecation failures

Hopefully this will let me debug the failing Travis test, since I
can't reproduce it locally..

njsmith committed:
MAINT: make the deprecation warnings a little more distinctive

This is probably a good change in any case, and in particular might
help with debugging whatever bizarro thing is going wrong on Travis.

njsmith commented:
I am utterly baffled by why the tests here are failing. Does anyone have any idea? The failing line is:
```
self.assert_deprecated(lambda: np.arange(2) != NotArray())
```
where `NotArray` is an object whose `__array__` method raises `TypeError`. On Python 3, but not Python 2, and only on Travis, not on my computer, then this fails because it apparently emits *two* copies of the deprecation warning -- first for a failed `!=` and then again for a failed `==` (!!).

I cannot for the life of me figure out why the above expression, on py3-on-travis, would involve any call to `==`. Note that Python does *not* convert `a != b` into `not a == b`, even if `a != b` gives a `NotImplemented`.

Any ideas...? I am tempted to just loosen the test so that this counts as a pass, but I feel like I will probably end up regretting that sooner or later...

mhvk commented:
Very odd.  FWIW, tests pass on my machine too (with python 3.4).

mhvk commented:
@njsmith - and of course fine from my side to turn array * list-of-strings into an error (or indeed apply the multiplier to each item). Amazing traceback...

charris commented:
Fails here with Python 3.4, I'll track it down.

charris commented:
```
>>> np.arange(2) != NotArray()
__main__:1: DeprecationWarning: elementwise != comparison failed; this will raise an error in the future.
__main__:1: DeprecationWarning: elementwise == comparison failed; this will raise an error in the future.
True
```
Apparently `!=` calls `==`.

njsmith commented:
Oh hooray you can reproduce it.

Yes != seems to be calling ==, but how and where and why? Python itself
(AFAICT) does not have any fallback from != to ==, and neither does numpy.
Except on some people's machines, apparently.

Could you stick a breakpoint on the line where that == warning is issued
and see if the stack trace reveals anything useful about who's calling it?
Or else I guess I'd try breaking on the != message and then single-stepping
to see where it goes from there.
On May 15, 2015 10:54 AM, "Charles Harris" <notifications@github.com> wrote:

> >>> np.arange(2) != NotArray()
> __main__:1: DeprecationWarning: elementwise != comparison failed; this will raise an error in the future.
> __main__:1: DeprecationWarning: elementwise == comparison failed; this will raise an error in the future.
> True
>
> Apparently != calls ==.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5864#issuecomment-102471076>.
>

charris commented:
I'm pretty sure it is this
```
        if (needs_right_binop_forward(obj_self, other, "__ne__", 0) &&
                Py_TYPE(obj_self)->tp_richcompare != Py_TYPE(other)->tp_richcompare) {
            Py_INCREF(Py_NotImplemented);
            return Py_NotImplemented;
        }
```
I suspect `>` and `>=` will have the same problem.

pv commented:
Yes, Python has a fallback to the reflected operation if the other one is not implemented.

njsmith commented:
Yes, but the reflected version of != is !=, not ==. That's not a reason why
it would be calling ==.
On May 15, 2015 11:39 AM, "Pauli Virtanen" <notifications@github.com> wrote:

> Yes, Python has a fallback to the reflected operation if the other one is
> not implemented.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5864#issuecomment-102488161>.
>

njsmith commented:
Also, that check @charris is pointing to comes *before* the != warning is
issued, so if that were it then we'd be returning early and no warning
would be issued.
On May 15, 2015 11:41 AM, "Nathaniel Smith" <njs@pobox.com> wrote:

> Yes, but the reflected version of != is !=, not ==. That's not a reason
> why it would be calling ==.
> On May 15, 2015 11:39 AM, "Pauli Virtanen" <notifications@github.com>
> wrote:
>
>> Yes, Python has a fallback to the reflected operation if the other one is
>> not implemented.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/numpy/numpy/pull/5864#issuecomment-102488161>.
>>
>

pv commented:
Indeed. Probably difficult to tell wth is going on without gdb :/

charris commented:
Note that  `!=` returns `NotImplemented`, python calls `==`, gets `NotImplemented`, and evidently has a third fallback position, which I'll guess is comparing memory locations.

njsmith commented:
As far as I can tell from both experimentation and reading the cpython
source, that is just not true. For 'a != b', python does the following
steps:
a.__ne__(b)
b.__ne__(a)
return True
What makes you think that python automatically calls ==?
On May 15, 2015 12:36 PM, "Charles Harris" <notifications@github.com> wrote:

> Note that != returns NotImplemented, python calls ==, gets NotImplemented,
> and evidently has a third fallback position, which I'll guess is comparing
> memory locations.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5864#issuecomment-102504074>.
>

charris commented:
The evidence is that it isn't implemented that way ;) I'll take a look.

charris commented:
It has something to do with the right hand class, Using just two arrays, `==` is never called.
```
>>> np.ones((3,2)) != np.ones((2,3))
__main__:1: DeprecationWarning: elementwise != comparison failed; this will raise an error in the future.
__main__:1: DeprecationWarning: elementwise != comparison failed; this will raise an error in the future.
True
```
It's the way errors with `__array__` are dealt with
```
>>> class Foo():
...     def __array__(self):
...       return np.ones(2,3)
... 
>>> np.ones((3,2)) != Foo()
__main__:1: DeprecationWarning: elementwise != comparison failed; this will raise an error in the future.
__main__:1: DeprecationWarning: elementwise == comparison failed; this will raise an error in the future.
```

njsmith commented:
Were you able to get a C stack trace for the second message? *Immediately* after the `!=` message is issued, `array_richcompare` returns `NotImplemented`, so your latest comment seems to suggest that something *that isn't numpy* is somehow calling `__array__`, which is just bizarre.

I can confirm though that on my machine, I just get one warning from your `Foo` example, so that suggests we're on the right track:
```
In [6]: np.ones((3, 2)) != Foo()
./runtests.py:1: DeprecationWarning: elementwise != comparison failed; this will raise an error in the future.
  #!/usr/bin/env python
Out[6]: True
```

njsmith commented:
Regarding the issue with `[1, 2] * np.array(2)` or `["a", "b"] * np.array(2)`: it looks like the answer is, all Python interpreters do in fact need special case code to handle these cases the way we want (e.g. PyPy has already implemented this, see the email thread on Python-dev), so it's de facto part of the language even if not de jure. So I think we can just declare that the correct behaviour for these cases is to perform elementwise multiplication, and as a first step towards that it's OK for this PR to make `["a", "b"] * np.array(2)` raise an error instead of returning `["a", "b", "a", "b"]`.

charris commented:
Oops, I notice that `np.ones(2,3)` raises an exception, class works correctly when fixed. Returning None also causes problems. So the problem seems to be a combination of `__array__` and `exception` or no return. Hmm. Anyway, I think you will always get two warnings with the `__arrray__` function, as the rhs will be an array, and both arrays will be called. The only places that I can see `__array__` being called suppress exceptions, so perhaps they return something else in that case.

njsmith commented:
Hi Chuck, this doesn't seem to be helping much... Any chance I could get a
shell on the box that reproduces the issue?
On May 15, 2015 3:09 PM, "Charles Harris" <notifications@github.com> wrote:

> Oops, I notice that np.ones(2,3) raises an exception, class works
> correctly when fixed. Returning None also causes problems. So the problem
> seems to be a combination of __array__ and exception or no return. Hmm.
> Anyway, I think you will always get two warnings with the __arrray__
> function, as the rhs will be an array, and both arrays will be called. The
> only places that I can see __array__ being called suppress exceptions, so
> perhaps they return something else in that case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5864#issuecomment-102535829>.
>

charris commented:
@njsmith Seems that any linux box with Python 3 will do, I'm running Fedora 21. I could give you an account on my home box and turn on ssh, although the address changes now and then, but I'd be surprised if you couldn't find one closer. I'm thinking Numpy is missing an error check somewhere...

charris commented:
Just in case, I assume you are running `python3 -Walways`.

charris commented:
However, I think you can fix the failing test without tracking down the source.
```
>>> class Foo():
...     def __array__(self):
...         return NotImplemented
...     def __ne__(self, other):
...         return NotImplemented
... 
>>> np.ones(3) != Foo()
__main__:1: DeprecationWarning: elementwise != comparison failed; this will raise an error in the future.
True
```
The comparison fails in the old way if `__ne__` is not defined.

charris commented:
Looks like the default is for `Foo` to implement `__ne__` by calling `__eq__` and returning the compliment.

njsmith commented:
I did try reproducing on Linux + python 3, and failed. That's what's so
weird...
.
That thing with __ne__ is very weird. I will follow up more, but note that
the python 3 docs explicitly say: "There are no implied relationships among
the comparison operators. The truth of x==y does not imply that x!=y is
false. Accordingly, when defining __eq__(), one should also define __ne__()
so that the operators will behave as expected." --
https://docs.python.org/3/reference/datamodel.html
On May 16, 2015 9:07 AM, "Charles Harris" <notifications@github.com> wrote:

> Looks like the default is for Foo to implement __ne__ by calling __eq__
> and returning the compliment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5864#issuecomment-102644081>.
>

charris commented:
>The truth of x==y does not imply that x!=y is false.

That doesn't mean that `__ne__` can't be implemented using `__eq__`. Perhaps "does not" should be replaced by "may not".

njsmith commented:
You're right, it looks like starting in Python 3.0, `object.__ne__` is implemented as
```
def __ne__(self, other):
    result = self.__eq__(other)
    if result is NotImplemented:
        return NotImplemented
    else:
        return not result
```
(This appears to be entirely undocumented except for a brief/vague reference in the "What's new in Python 3.0" doc, and the reason I missed it before is that I was looking at the generic dispatch logic for comparisons, while this is specifically inside the implementation of `object`.)

This still leaves the mystery of why Python 3 only seems to do this on your machine and Travis-CI, but not on my machine or @mhvk's...

charris commented:
This needs to be finished up for 1.10.0. One thing I want to settle is what to do for `@`. At the moment it can be overridden by defining both `__ufunc_override__` and `__rmatmul__`, but that isn't backward compatible. For backward compatibility I think it needs to also handle `__array_priority__` together with `__rmatmul__`. Once this goes in the tests need to be updated.

charris commented:
@njsmith At this point I'm inclined to put this off until 1.11.

mhvk commented:
It would seem goot to get this in; as I understand it, the issue with the double DeprecationWarning is now understood to be unrelated (if still odd).

charris commented:
Well, 1) it's unfinished and 2) we haven't come to a conclusion on the topic. Current numpy isn't broken, just not what it will be. I'd like to be in a position to branch 1.10 in a week.

njsmith commented:
This pr is totally orthogonal to the __numpy_ufunc__ discussion. There is
no controversy here (AFAIK!), it just needs a bit of work on the tests.
(And I've been a bit slow because of the __numpy_ufunc__ discussion taking
my attention.) This pr is not a 1.10 blocker, but it would be nice to get
it in. I can finish it up this week, no problem. The __numpy_ufunc__
discussion is a 1.10 blocker, because current master is not in a releasable
state wrt __numpy_ufunc__.
On Jun 2, 2015 2:28 PM, "Charles Harris" <notifications@github.com> wrote:

> Well, 1) it's unfinished and 2) we haven't come to a conclusion on the
> topic. Current numpy isn't broken, just not what it will be. I'd like to be
> in a position to branch 1.10 in a week.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5864#issuecomment-108103600>.
>

charris commented:
@njsmith So what breaks with current master?

njsmith commented:
The problem with current master is that it contains `__numpy_ufunc__`
support, but it's not paired with any coherent plan for how to handle
the interaction between `__numpy_ufunc__` and binops. So if we release
master as-is, we'll end up locking ourselves into some unconsidered
resolution to #5844. Our options are to either pick a resolution (I'll
make an attempt at getting that discussion to at least a
good-enough-for-1.10 resolution this week), or else we can disable
`__numpy_ufunc__` support in 1.10 like we did for 1.9. I'd really
prefer the former, but if we can't get things resolved in time then
waiting for 1.11 is better than saddling ourselves with yet another
backcompat millstone.

On Tue, Jun 2, 2015 at 3:00 PM, Charles Harris <notifications@github.com> wrote:
> @njsmith So what breaks with current master?
>
> —
> Reply to this email directly or view it on GitHub.



-- 
Nathaniel J. Smith -- http://vorpus.org

charris commented:
How does it lock us in?

njsmith commented:
People start using it -> we can't change it without breaking backcompat.

On Tue, Jun 2, 2015 at 3:16 PM, Charles Harris <notifications@github.com> wrote:
>
> How does it lock us in?
>
> —
> Reply to this email directly or view it on GitHub.




-- 
Nathaniel J. Smith -- http://vorpus.org

charris commented:
What specific usage will break?

charris commented:
@pv Are you OK with this?

njsmith commented:
My current todo list for this PR, to aid in evaluating this:
* Fix the failing test
* Add the tests that @pv [requested](https://github.com/numpy/numpy/pull/5864#issuecomment-101194602)
* ...maybe something about in-place arguments? @pv: If you could reply to [my comment above](https://github.com/numpy/numpy/pull/5864#issuecomment-102215153) that would be keen.

charris commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1351,6 +1369,11 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
             if (array_other == NULL) {
                 PyErr_Clear();
+                if (DEPRECATE(
+                        "elementwise == comparison failed and returning scalar "
```
Might be clearer if the text lines up as an argument to `DEPRECATE(`

charris commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1359,18 +1382,31 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
             if (_res == 0) {
-                Py_DECREF(result);
                 Py_DECREF(array_other);
+                if (DEPRECATE_FUTUREWARNING(
```
But that won't work here. Hmm....

charris commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -896,16 +864,122 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+            return -2;
+        }
+        else if (!strcmp(ufunc_name, "less")
+                 || !strcmp(ufunc_name, "less_equal")
```
Style: put `||` at ends of previous lines.

charris replied:
Not to say I didn't argue for this style once, but it is not the way of Python.

charris commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -4222,13 +4255,15 @@ ufunc_generic_call(PyUFuncObject *ufunc, PyObject *args, PyObject *kwds)
         }
         else if (ufunc->nin == 2 && ufunc->nout == 1) {
-            /* To allow the other argument to be given a chance */
+            /* For array_richcompare's benefit -- see the long comment in
```
Multiline comments like
```
/*
 * blah
 */
```

charris commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -4222,13 +4255,15 @@ ufunc_generic_call(PyUFuncObject *ufunc, PyObject *args, PyObject *kwds)
         else {
             PyErr_SetString(PyExc_TypeError,
-                                        "Not implemented for this type");
+                            "XX can't happen, please report a bug XX");
```
Don't forget the `XX`.

charris commented on numpy/core/tests/test_deprecations.py:
```diff
@@ -76,15 +77,25 @@ def assert_deprecated(self, function, num=1, ignore_others=False,
-            raise AssertionError("%i warnings found but %i expected"
-                                                        % (len(self.log), num))
+            raise AssertionError("%i warnings found but %i expected:\n"
+                                 % (len(self.log), num)
```
Operators at end of line would probably be better.

charris commented on numpy/core/tests/test_deprecations.py:
```diff
@@ -442,6 +452,52 @@ def test_scalar_none_comparison(self):
+                        with warnings.catch_warnings() as l:
+                            warnings.filterwarnings("always")
+                            assert_raises(TypeError, f, arg1, arg2)
+                            assert not l
```
Debugging?

charris commented:
OK, this doesn't bother me. I'll update the `@` stuff to match when this goes in.

charris commented:
Might want to make a quick review of the C style guide.

charris commented on numpy/ma/core.py:
```diff
@@ -3678,6 +3678,16 @@ def __repr__(self):
             return _print_templates['short_std'] % parameters
         return _print_templates['long_std'] % parameters
 
+    def _delegate_binop(self, other):
```
This is an improvement, but the fact that it is needed worries me. What other applications are going to break because of this change?

charris replied:
Will masked array still work without these changes, maybe with some warnings?

charris replied:
Might want to remove this bit up above also
```
        # check it worked
        if result is NotImplemented:
            return NotImplemented
```

charris commented:
@njsmith Needs finishing.

mhvk commented:
@njsmith - can I help in some way? I like the PR both generally and would like to rebase some `ma` stuff after this gets in (notably #3907).

charris commented:
@njsmith I will need to leave this out of the 1.10 release if it isn't finished up this weekend. It is blocking progress on other PRs.

charris commented:
Cleaned up in #5964, so closing this.
